pkg/util/kernel/headers.go

```diff
@@ -0,0 +1,157 @@
+// +build linux
+
+package kernel
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+
+	"github.com/DataDog/datadog-agent/pkg/metadata/host"
+	"github.com/mholt/archiver/v3"
+)
+
+const sysfsHeadersPath = "/sys/kernel/kheaders.tar.xz"
+const kernelModulesPath = "/lib/modules/%s/build"
+
+func FindHeaderDirs() ([]string, error) {
+	hv, err := HostVersion()
+	if err != nil {
+		return nil, fmt.Errorf("unable to determine host kernel version: %w", err)
+	}
+
+	dir, err := getHeaderDirs(hv)
+	if err == nil {
+		return dir, nil
+	}
+
+	if os.IsNotExist(err) {
+		// as last resort, look for sysfs headers
+		if dir, syserr := getSysfsHeaderDirs(hv); syserr == nil {
+			return dir, nil
+		}
+	}
+	return nil, err
+}
+
+func getHeaderDirs(v Version) ([]string, error) {
+	hi := host.GetStatusInformation()
+	if hi.KernelVersion == "" {
+		return nil, fmt.Errorf("unable to get host metadata")
+	}
+
+	dirs := []string{kernelModulesPath}
+	if hi.Platform == "debian" {
+		dirs = append(dirs, "/lib/modules/%s/source")
+	}
+	// KernelVersion == uname -r
+	for i, _ := range dirs {
+		dirs[i] = fmt.Sprintf(dirs[i], hi.KernelVersion)
+	}
+
+	for _, d := range dirs {
+		hv, err := getHeaderVersion(d)
+		if err != nil {
+			// if version.h is not found in this directory, keep going
+			if os.IsNotExist(err) {
+				continue
+			}
+			return nil, fmt.Errorf("error validating headers version: %w", err)
+		}
+		if hv == v {
+			return dirs, nil
+		}
+	}
+	return nil, os.ErrNotExist
+}
+
+func getHeaderVersion(path string) (Version, error) {
+	vh := filepath.Join(path, "include/generated/uapi/linux/version.h")
+	f, err := os.Open(vh)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+	return parseHeaderVersion(f)
+}
+
+func parseHeaderVersion(r io.Reader) (Version, error) {
+	re := regexp.MustCompile(`^#define[\t ]+LINUX_VERSION_CODE[\t ]+(\d+)$`)
```

Comment:
nit: this function gets called a lot, so this regexp should be pulled out to a const